src/spox/_value_prop_backend.py

```diff
@@ -0,0 +1,123 @@
+# Copyright (c) QuantCo 2023-2025
+# SPDX-License-Identifier: BSD-3-Clause
+from __future__ import annotations
+
+import logging
+from abc import ABC, abstractmethod
+from collections.abc import Iterator
+from contextlib import contextmanager
+from typing import TYPE_CHECKING, Any, Generic, TypeVar
+
+import onnx
+
+from . import _value_prop
+from ._type_system import Type
+from ._value_prop import ORTValue, PropValue, RefValue
+
+if TYPE_CHECKING:
+    pass
+
+TValue = TypeVar("TValue")
+
+
+class BaseValuePropBackend(ABC, Generic[TValue]):
+    @abstractmethod
+    def wrap_feed(self, value: PropValue) -> TValue:
+        raise NotImplementedError
+
+    @abstractmethod
+    def run(
+        self, model: onnx.ModelProto, input_feed: dict[str, TValue]
+    ) -> dict[str, TValue]:
+        raise NotImplementedError
+
+    @abstractmethod
+    def unwrap_feed(self, typ: Type, value: TValue) -> PropValue:
+        raise NotImplementedError
+
+
+class ReferenceValuePropBackend(BaseValuePropBackend[RefValue]):
+    def wrap_feed(self, value: PropValue) -> RefValue:
+        return value.to_ref_value()
+
+    def run(
+        self, model: onnx.ModelProto, input_feed: dict[str, RefValue]
+    ) -> dict[str, RefValue]:
+        try:
+            session = onnx.reference.ReferenceEvaluator(model)
+            output_feed = dict(zip(session.output_names, session.run(None, input_feed)))
+        except Exception as e:
+            # Give up on value propagation if an implementation is missing.
+            logging.debug(
+                f"Value propagation in {model} on the ONNX reference implementation failed with - "
+                f"{type(e).__name__}: {e}"
+            )
+            return {}
+        return output_feed
+
+    def unwrap_feed(self, typ: Type, value: RefValue) -> PropValue:
+        return PropValue.from_ref_value(typ, value)
+
+
+class OnnxruntimeValuePropBackend(BaseValuePropBackend[ORTValue]):
+    def __init__(self, session_options: dict[str, Any] | None = None) -> None:
```

Comment:
I don't think we need the `session_options` field here. It isn't used within Spox and as for custom operators, the ABC is sufficient.